contracts/components/node_runners/NodeRunnerRegistryCore.sol

```diff
@@ -0,0 +1,379 @@
+import "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721EnumerableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/cryptography/draft-EIP712Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/cryptography/SignatureCheckerUpgradeable.sol";
+
+import "hardhat/console.sol";
+
+abstract contract NodeRunnerRegistryCore is BaseComponentUpgradeable, ERC721Upgradeable, ERC721EnumerableUpgradeable, StakeSubjectUpgradeable, EIP712Upgradeable {
+    using CountersUpgradeable for CountersUpgradeable.Counter;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    struct ScannerNode {
+        bool registered;
+        bool disabled;
+        uint256 nodeRunnerId;
+        uint256 chainId;
+        string metadata;
+    }
+    struct ScannerNodeRegistration {
+        address scanner;
+        uint256 nodeRunnerId;
+        uint256 chainId;
+        string metadata;
+        uint256 timestamp;
+    }
+
+    bytes32 private constant _SCANNERNODEREGISTRATION_TYPEHASH =
+        keccak256("ScannerNodeRegistration(address scanner,uint256 nodeRunnerId,uint256 chainId,string metadata,uint256 timestamp)");
+
+    /// nodeRunnerIds is a sequential autoincremented uint
+    CountersUpgradeable.Counter private _nodeRunnerIdCounter;
+    /// ScannerNode data for each scanner address;
+    mapping(address => ScannerNode) _scannerNodes;
+    /// Set of Scanner Node addresses each nodeRunnerId owns;
+    mapping(uint256 => EnumerableSet.AddressSet) internal _scannerNodeOwnership;
+    /// StakeThreshold of each chainId;
+    mapping(uint256 => StakeThreshold) internal _stakeThresholds;
+    /// amount of time allowed from scanner signing a ScannerNodeRegistration and it's execution by NodeRunner
+    uint256 public registrationDelay;
+
+    event ScannerUpdated(uint256 indexed scannerId, uint256 indexed chainId, string metadata, uint256 nodeRunner);
+    event StakeThresholdChanged(uint256 indexed chainId, uint256 min, uint256 max, bool activated);
+    event RegistrationDelaySet(uint256 delay);
+    event ScannerEnabled(uint256 indexed scannerId, bool indexed enabled, address sender, bool disableFlag);
+
+    error NodeRunnerNotRegistered(uint256 nodeRunnerId);
+    error ScannerExists(address scanner);
+    error ScannerNotRegistered(address scanner);
+    error ScannerAlreadyRegisteredTo(address scanner, uint256 nodeRunnerId);
+    error ScannerNotRegisteredTo(address scanner, uint256 nodeRunnerId);
+    error PublicRegistrationDisabled(uint256 chainId);
+    error RegisteringTooLate();
+    error SignatureDoesNotMatch();
+    error CannotSetScannerActivation();
+
+    /**
+     * @notice Checks sender (or metatx signer) is owner of the scanner token.
+     * @param nodeRunnerId ERC721 token id of the Node Runner.
+     */
+    modifier onlyOwnerOf(uint256 nodeRunnerId) {
```

Comment:
I found this concept (owner of a node runner ID) particularly confusing, it's not clear that the owner _is_ the node runner. From the documentation it seems we are just refering to the account as "node runner" (and to the token as "node runner ID").

I would rename this modifier to `onlyNodeRunner(nodeRunnerId)`.

We need to be particularly clear around this because there is potential for confusion around the node runner _account_ / node runner _ID_ distinction.

Reply:
Hmm agree, the concept of ownership of the ERC721 leaks here and clashes with ownership of the Scanner Nodes.

Changed it: 7cec852